src/components/ViewedByCustomer/ViewedByCustomer.ts

```diff
@@ -0,0 +1,65 @@
+import { Component, ComponentOptions, IComponentBindings, $$, Initialization } from 'coveo-search-ui';
+import { user } from '../../utils/icons';
+
+/**
+ * The options for the ViewedByCustomerComponent
+ */
+export interface IViewedByCustomerOptions {
+    showIcon?: boolean;
+    label?: string;
+}
+
+/**
+ * Result component indicating if a search result have been viewed by the targeted user.
+ */
+export class ViewedByCustomer extends Component {
+    /**
+     * Unique Identifier used by the Search-UI.
+     */
+    public static readonly ID = 'ViewedByCustomer';
+
+    /**
+     * Default options used by the component.
+     */
+    public static readonly options: IViewedByCustomerOptions = {
+        /**
+         * If true, will display an icon when the component is displayed itself.
+         */
+        showIcon: ComponentOptions.buildBooleanOption({ defaultValue: true }),
+        /**
+         * The label that should be displayed when the component is displayed.
+         */
+        label: ComponentOptions.buildStringOption({ defaultValue: 'Viewed by Customer' })
+    };
+
+    /**
+     * Create an instance of {@link ViewedByCustomer}.
+     * @param element Element on which to bind the component.
+     * @param options Initialization options of the component.
+     * @param bindings Bindings of the Search-UI environment.
+     */
+    public constructor(public element: HTMLElement, public options: IViewedByCustomerOptions, public bindings: IComponentBindings) {
+        super(element, ViewedByCustomer.ID, bindings)/* istanbul ignore next Issue with Istanbul and super calls*/;
+        this.options = ComponentOptions.initComponentOptions(element, ViewedByCustomer, options);
+
+        if (this.resolveResult().isUserActionView) {

```

Comment:
is resolveResult safe?
Will this assert if the result is null for example or something like that?

Reply:
There's an assert :

[Component.ts](https://github.com/coveo/search-ui/blob/master/src/ui/Base/Component.ts#L363)
```typescript
  public resolveResult(): IQueryResult {
    return Component.getResult(this.element);
  }
```
[Component.ts](https://github.com/coveo/search-ui/blob/master/src/ui/Base/Component.ts#L408)
```typescript
  static getResult(element: HTMLElement, noThrow: boolean = false): IQueryResult {
    var resultElement = $$(element).closest('.CoveoResult');
    Assert.check(noThrow || resultElement != undefined);
    return resultElement['CoveoResult'];
  }```

We look up for the first result by ascending the DOM tree from the Viewed By Customer element, if we do not found a result, being a component that should be used on a ResultTemplate, that's a legitimate usage issue and the Assert will throw rightfully so.